src/ui_manager.cpp

```diff
@@ -421,14 +454,25 @@ void ui_adaptor::redraw_invalidated()
                         ui.invalidated = false;
                     }
                 }
+                if( ui.is_imgui == draw_imgui && ( !ui.is_imgui || imgui_is_on_top ) ) {
```

Comment:
if draw_imgui is true, only draw imgui windows. if false, we draw non-imgui windows. There probably isn't a case where a non-imgui window will use the inheritance-based redraw() function, so maybe this can be changed.

ImGui windows defy our invalidation logic - it tracks internally what needs to be redrawn, so we must redraw the window every frame. Occlusion is all handled by the library. We will need a different approach to handle screen readers.

Reply:
> if draw_imgui is true, only draw imgui windows. if false, we draw non-imgui windows.

I mean, when `draw_imgui` is false and `ui.is_imgui` is false, `ui.redraw()` is called, which is not implemented for non-imgui `ui_adptor`s, which is why I don't understand the logic here.

> it tracks internally what needs to be redrawn, so we must redraw the window every frame.

I see. In that case, I suggest you set `invalidated` of imgui windows to true before redraw, call the imgui redraw function in place of redraw_cb, and also set the window size in the resize callback, so the occlusion check is done correctly and the invariants of invalidate flag optimization are preserved. I think this way the logic is simpler and easier to understand.

> Occlusion is all handled by the library.

My concern is how it interacts with non-imgui windows when both are present on-screen. The occlusion check of `ui_adaptor` has some tricky assumptions and this might break it if not handled carefully. For example, suppose A is a imgui window and B and C are non-imgui windows. When the code first creates C and calls `ui_manager::redraw` to invalidate C and redraw, the occlusion check and invalidated flag optimization will decide that A and B are both already on-screen and thus do not need to be redrawn. If you always redraw A but do not se the `invalidated` flag of A nor the size of A, the occlusion check and invalidated flag optimization will think that B is not occluded by the redrawn A, and will not redraw B as a result, so in the end you end up with B missing from the screen.

I haven't tested the above scenario yet, but my point is that it's easier for you to integrate the new UI into the existing logic rather than complicating the logic (and even break it as a result).

```
+------------+
|A +------+  |  +------+
|  |B     |  |  |C     |
|  +------+  |  +------+
+------------+
```

> We will need a different approach to handle screen readers.

Could you elaborate? Screen readers are not currently supported on the tiles build so it's fine if ImGUI window does not support it, but I think ImTUI window can still support it. Currently screen reader compatibility is handled by placing the terminal cursor at the correct position after redraw completes, and I imagine a ImTUI window can set the cursor position via `ui_adaptor` similar to how non-imgui windows do it.

Reply:
>I mean, when draw_imgui and ui.is_imgui are both false, ui.redraw() is called, which is not implemented for non-imgui ui_adptors, which is why I don't understand the logic here. Edit: also, if ui.is_imgui and draw_imgui are both true and the topmost UI is not a imgui window, then no imgui window will be redrawn, which also seems weird to me.

right, that's why i said i will probably change that if statement. i just havent yet

to your last point, and this is to do with occlusion - ImGui seems to always draw all of its windows on a single layer, no matter what I do. So either it's on top of everything, or it isn't. So the solution I came up with is that, when a non-ImGui ui_adaptor is placed in the stack above an ImGui window, I just skip drawing all ImGui windows to avoid any collisions.

>Could you elaborate? Screen readers are not currently supported on the tiles build so it's fine if ImGUI window does not support it, but I think ImTUI window can still support it. Currently screen reader compatibility is handled by placing the terminal cursor at the correct position after redraw completes, and I imagine a ImTUI window can set the cursor position via ui_adaptor similar to how non-imgui windows do it.

I really don't know one way or another what will happen, I just think that it's a safe bet to say things will break. Even in ncurses builds I need to redraw constantly, whether input is received or not, in order for ImTui to properly react to input and draw. It's just the way ImGui works. You could try and load up my branch in an ncurses build and try the screen reader and see what happens?

Reply:
> ImGui seems to always draw all of its windows on a single layer, no matter what I do.

I've never used ImGui so I can't help much, but could its output be redirected to some frame buffer so you can copy its content to the SDL frame buffer when needed? For ImTui, could it be instructed to output to a particular curses window without refreshing? I think that might solve the layering problem.

Or you could do it in reverse by treating `ui_adaptor`s as layers in the ImGui system, but I don't know how ImGui handles layers so I'm not sure if how easy that would be.

> Even in ncurses builds I need to redraw constantly, whether input is received or not, in order for ImTui to properly react to input and draw.

Constant redrawing has been reported previously to greatly increase the CPU consumption in some terminals so I'm afraid that would be considered a regression.

> You could try and load up my branch in an ncurses build and try the screen reader and see what happens?

Sure, I'll test it later.

Reply:
>Constant redrawing has been reported previously to greatly increase the CPU consumption in some terminals so I'm afraid that would be considered a regression.

So, when you make a call to ImTui to render, it will only write things to stdscr when an element of the screen changes. Even though you call the redraw() function and the ImGui::Begin and ImGui::End calls are sent, ImGui keeps internal buffers of what is currently on the screen, and only redraws the portions of the screen that have changed. I could have the ImTui draw calls return a value if anything has been drawn to the screen, and only call doupdate when it is true. We would still need to change our input handling so that doupdate is either not called there, or only called when necessary (i.e. when some input is received or when ImTui is active and ImTui has indicated that it drew stuff)

>I've never used ImGui so I can't help much, but could its output be redirected to some frame buffer so you can copy its content to the SDL frame buffer when needed? For ImTui, could it be instructed to output to a particular curses window without refreshing? I think that might solve the layering problem.

We could maybe try a method where the screen does not update unless ImGui indicates it has drawn something. I'm not sure whether there would be issues raised because of this. The drawing "strategy" was copied right from ImGui's SDL renderer demos so I am unsure how it will function if the order of operations will change.

Reply:
From your other comment:

> The ImTui calls do not update the screen, they write to stdscr

If it only writes to stdscr and do not call `wnoutrefresh` nor `wrefresh`, you can try calling `wnoutrefresh( stdscr )` after each imgui window is redrawn, so that the updates are copied to the frame buffer but not immediately displayed. And IIRC, only the updates after the previous refresh will be copied to the frame buffer so that could solve the problem of imgui drawing all windows on a single layer.

As for SDL, even if it does not support using a separate frame buffer, I think you can work it around by setting a new frame buffer in SDL, draw the window, and then swap out the frame buffer to copy the relevant part to the original frame buffer.

Reply:
@Qrox are you satisfied with where this code is at? I could take another crack at this, maybe I could have code which just marks imgui windows as invalid on every redraw? I'm not sure where else to go with this code.